logbook-json/src/main/java/org/zalando/logbook/json/JsonHttpLogFormatter.java

```diff
@@ -1,92 +1,191 @@
+import org.zalando.logbook.Origin;
+import org.zalando.logbook.Precorrelation;
+
+import com.fasterxml.jackson.core.JsonFactory;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.core.io.JsonStringEncoder;
+import com.fasterxml.jackson.core.util.BufferRecyclers;
+import com.fasterxml.jackson.databind.ObjectMapper;
 
 /**
- * A custom {@link HttpLogFormatter} that produces JSON objects. It can be augmented with composition:
- *
- * <pre>
- * {@code
- *
- * public class CustomsFormatter implements HttpLogFormatter {
- *
- *     private final JsonHttpLogFormatter delegate;
- *
- *     public CustomsFormatter(final ObjectMapper mapper) {
- *         this.delegate = new JsonHttpLogFormatter(mapper);
- *     }
- *
- *     public String format(final Precorrelation<HttpRequest> precorrelation) throws IOException {
- *         Map<String, Object> request = delegate.prepare(precorrelation);
- *         // modify request here
- *         return delegate.format(request);
- *     }
- *
- *     public String format(final Correlation<HttpRequest, HttpResponse> correlation) throws IOException {
- *         Map<String, Object> response = delegate.prepare(correlation);
- *         // modify response here
- *         return delegate.format(response);
- *      }
- *
- * }
- * }
- * </pre>
+ * A custom {@link HttpLogFormatter} that produces JSON objects. 
  */
 @API(status = STABLE)
-public final class JsonHttpLogFormatter implements StructuredHttpLogFormatter {
-
-    private final ObjectMapper mapper;
+public final class JsonHttpLogFormatter implements HttpLogFormatter {
 
+    private final JsonFactory jsonFactory;
+    
     public JsonHttpLogFormatter() {
         this(new ObjectMapper());
     }
+    
+    public JsonHttpLogFormatter(JsonFactory jsonFactory) {
+        this.jsonFactory = jsonFactory;
+    }
 
-    public JsonHttpLogFormatter(final ObjectMapper mapper) {
-        this.mapper = mapper;
+    public JsonHttpLogFormatter(ObjectMapper objectMapper) {
+        this.jsonFactory = objectMapper.getFactory();
     }
 
     @Override
-    public Optional<Object> prepareBody(final HttpMessage message) throws IOException {
-        final String contentType = message.getContentType();
-        final String body = message.getBodyAsString();
-        if(body.isEmpty()) {
-            return Optional.empty();
+    public String format(Precorrelation precorrelation, HttpRequest request) throws IOException {
+        final String correlationId = precorrelation.getId();
+        
+        String body = request.getBodyAsString();
+
+        StringBuilderWriter writer = new StringBuilderWriter(body.length() + 2048);
+
+        JsonGenerator generator = jsonFactory.createGenerator(writer);
+        try {
+            generator.writeStartObject();
+            
+            generator.writeFieldName("origin");
+            if(request.getOrigin() == Origin.LOCAL) {
+                generator.writeRawValue("\"local\"");
```

Comment:
Why not `writeString("local")`?

Reply:
Marginally faster; no escaping necessary.